src/librbd/api/Group.cc

```diff
@@ -318,80 +628,273 @@ int Group<I>::image_list(librados::IoCtx& group_ioctx,
+	  ne << dendl;
+	ret_code = -ENOENT;
+      } else {
+	image_snaps[i].snap_id = snapid_t(snap_id);
+	image_snaps[i].pool = ictx->data_ctx.get_id();
+	image_snaps[i].image_id = ictx->id;
+      }
+    }
+  }
+  if (ret_code != 0) {
+    goto remove_image_snaps;
+  }
+
+  group_snap.snaps = image_snaps;
+  group_snap.state = cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE;
+
+  r = cls_client::group_snap_update(&group_ioctx, group_header_oid, group_snap);
+  if (r < 0) {
+    ret_code = r;
+    goto remove_image_snaps;
+  }
+
+  goto finish;
+
+remove_image_snaps:
+
+  for (int i = 0; i < image_count; ++i) {
+    ImageCtx *ictx = ictxs[i];
+    ldout(cct, 20) << "Removing individual snapshot with name: " <<
+      ind_snap_name << dendl;
+
+    on_finishes[i] = new C_SaferCond;
+    std::string snap_name;
+    ictx->snap_lock.get_read();
+    snap_t snap_id = ictx->get_snap_id_from_namespace(ne);
+    r = ictx->get_snap_name(snap_id, &snap_name);
+    ictx->snap_lock.put_read();
+    if (r >= 0) {
+      ictx->operations->snap_remove(ne, snap_name.c_str(), on_finishes[i]);
+    }
+    // Ignore missing image snapshots. The whole snapshot could have been inconsistent.
+  }
+
+  for (int i = 0, n = on_finishes.size(); i < n; ++i) {
+    r = on_finishes[i]->wait();
+    delete on_finishes[i];
+    if (r < 0 && r != -ENOENT) { // if previous attempts to remove this snapshot failed then the image's snapshot may not exist
+      lderr(cct) << "Failed cleaning up image snapshot. Ret code: " << r << dendl;
+      // just report error, but don't abort the process
+    }
+  }
+
+remove_record:
+  r = cls_client::group_snap_remove(&group_ioctx, group_header_oid,
+      group_snap.id);
+  if (r < 0) {
+    lderr(cct) << "error while cleaning up group snapshot" << dendl;
+    // we ignore return value in clean up
+  }
+
+finish:
+  for (int i = 0, n = ictxs.size(); i < n; ++i) {
+    if (ictxs[i] != nullptr) {
+      ictxs[i]->state->close();
+    }
+  }
+  return ret_code;
 }
 
 template <typename I>
-int Group<I>::image_get_group(I *ictx, group_spec_t *group_spec)
+int Group<I>::snap_remove(librados::IoCtx& group_ioctx, const char *group_name,
+    const char *snap_name)
 {
-  int r = ictx->state->refresh_if_required();
-  if (r < 0)
+  CephContext *cct = (CephContext *)group_ioctx.cct();
+  librados::Rados rados(group_ioctx);
+
+  std::vector<cls::rbd::GroupSnapshot> snaps;
+  std::vector<C_SaferCond*> on_finishes;
+
+  string group_id;
+  int r = cls_client::dir_get_id(&group_ioctx, RBD_GROUP_DIRECTORY,
+				 group_name, &group_id);
+  if (r < 0) {
+    lderr(cct) << "error reading consistency group id object: "
+	       << cpp_strerror(r)
+	       << dendl;
     return r;
+  }
+  string group_header_oid = util::group_header_name(group_id);
 
-  if (-1 != ictx->group_spec.pool_id) {
-    librados::Rados rados(ictx->md_ctx);
-    IoCtx ioctx;
-    rados.ioctx_create2(ictx->group_spec.pool_id, ioctx);
+  r = group_snap_list(group_ioctx, group_name, &snaps);
```

Comment:
Good use for a ```group_snap_get_by_name``` function. That would allow the removal of cls_rbd::check_duplicate_snap_name because it can just use the a helper "get_by_name" function